src/celengine/stardb.cpp

```diff
@@ -47,107 +54,171 @@ constexpr const char LacailleCatalogPrefix[]  = "Lacaille ";
+
+    const char* const end_ptr = name.data() + name.size();
+
+    std::array<AstroCatalog::IndexNumber, 3> tycParts;
+    auto result = from_chars(name.data() + pos, end_ptr, tycParts[0]);
+    if (result.ec != std::errc{}
+        || tycParts[0] < TYC123_MIN || tycParts[0] > TYC1_MAX
+        || result.ptr == end_ptr
+        || *result.ptr != '-')
+    {
+        return false;
+    }
+
+    result = from_chars(result.ptr + 1, end_ptr, tycParts[1]);
+    if (result.ec != std::errc{}
+        || tycParts[1] < TYC123_MIN || tycParts[1] > TYC2_MAX
+        || result.ptr == end_ptr
+        || *result.ptr != '-')
+    {
+        return false;
+    }
+
+    if (result = from_chars(result.ptr + 1, end_ptr, tycParts[2]);
+        result.ec == std::errc{}
+        && tycParts[2] >= TYC123_MIN
+        && (tycParts[2] <= TYC3_MAX
+            || (tycParts[2] == TDSC_TYC3_MAX && tycParts[0] <= TDSC_TYC3_MAX_RANGE_TYC1)))
+    {
+        // Do not match if suffix is present
+        pos = name.find_first_not_of(" \t", result.ptr - name.data());
+        if (pos != std::string_view::npos) { return false; }
+
+        catalogNumber = tycParts[2] * TYC3_MULTIPLIER
+                      + tycParts[1] * TYC2_MULTIPLIER
+                      + tycParts[0];
+        return true;
     }
 
     return false;
 }
 
 
-static bool parseCelestiaCatalogNumber(const std::string& name,
-                                       AstroCatalog::IndexNumber* catalogNumber)
+bool parseCelestiaCatalogNumber(std::string_view name,
+                                AstroCatalog::IndexNumber& catalogNumber)
 {
-    char extra[4];
-
-    if (name[0] == '#')
+    using celestia::compat::from_chars;
+    if (name.size() == 0 || name[0] != '#') { return false; }
+    if (auto [ptr, ec] = from_chars(name.data() + 1, name.data() + name.size(), catalogNumber); ec == std::errc{})
     {
-        unsigned int num;
-        if (sscanf(name.c_str(), "#%u %c", &num, extra) == 1)
-        {
-            *catalogNumber = (AstroCatalog::IndexNumber) num;
-            return true;
-        }
+        // Do not match if suffix is present
+        auto pos = name.find_first_not_of(" \t", ptr - name.data());
+        return pos == std::string_view::npos;
     }
 
     return false;
 }
 
 
+std::string catalogNumberToString(AstroCatalog::IndexNumber catalogNumber)
+{
+    if (catalogNumber <= StarDatabase::MAX_HIPPARCOS_NUMBER)
+    {
+        return fmt::format("HIP {}", catalogNumber);
+    }
+    else
+    {
+        AstroCatalog::IndexNumber tyc3 = catalogNumber / TYC3_MULTIPLIER;
+        catalogNumber -= tyc3 * TYC3_MULTIPLIER;
+        AstroCatalog::IndexNumber tyc2 = catalogNumber / TYC2_MULTIPLIER;
+        catalogNumber -= tyc2 * TYC2_MULTIPLIER;
+        AstroCatalog::IndexNumber tyc1 = catalogNumber;
+        return fmt::format("TYC {}-{}-{}", tyc1, tyc2, tyc3);
+    }
+}
+
+
+void stcError(const Tokenizer& tok,
+              const std::string& msg)
```

Comment:
could you check if it makes sense to use std::string_view here as well?

Reply:
done